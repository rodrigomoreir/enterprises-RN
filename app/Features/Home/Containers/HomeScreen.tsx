import React, { useEffect, useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import { 
  View, 
  KeyboardAvoidingView, 
  Platform, 
  ScrollView,
  ActivityIndicator,
  Dimensions
} from 'react-native';
import { getStyles } from './HomeScreen.style'
import { Card, Filter } from '../Components'
import { Header } from '../../../Components'
import { TextField } from '../../../Components'
import { Form } from '@unform/mobile'
import api from '../../../Services/api';
import { useAuth } from '../../../Hooks/Auth'
import { EnterpriseProps } from '../Entities'

const HomeScreen: React.FC = () => {
  const { client, token, uid } = useAuth()
  const navigation = useNavigation()
  const styles = getStyles()
  const [enterprises, setEnterprises] = useState<EnterpriseProps[]>([])
  const [loading, setLoading] = useState(true)

  const navigateToEnterpriseDetails = (enterprise: EnterpriseProps) => {
    navigation.navigate('EnterpriseDetails', { enterprise })
  }
  const navigateToUserProfile = () => navigation.navigate('UserProfile')

  useEffect(() => {
    console.log(token, client, uid)
    console.log('executou')
    const getEnterprises = async () => {
      await api.get('enterprises', {
        headers: { 'access-token': token, 'client': client, 'uid': uid }
      }).then(response => {
        setEnterprises(response.data.enterprises)
        setLoading(false)
      }).catch(error => {
        console.log(error)
      })
    }
    getEnterprises()
  }, []) 

  return (
    <KeyboardAvoidingView
      style={{ flex: 1 }}
      behavior={Platform.OS === 'ios' ? 'padding' : undefined}
      enabled
    >
      <Header title={'Empresas'} goProfile={navigateToUserProfile} />
      <Form onSubmit={() => {}} style={{flexDirection: 'row', }}>
        <TextField 
          name={'Pesquisar'} 
          style={styles.searchInput}
          buttonSearch={() => {}}
          placeholder={'Pesquisar'} 
          autoCorrect={false}
          onSubmitEditing={() => {
          }}
        />
          <Filter 
            filterId={enterprises.map(item => item.enterprise_type.id)} 
            filterName={enterprises.map(item=>item.enterprise_type.enterprise_type_name)}  
          />
      </Form>
      
      <ScrollView
        keyboardShouldPersistTaps='handled'
        style={styles.containerScrollView}
        contentContainerStyle={styles.contentScrollView}
      >
        {loading && (
          <View style={{flex:1, alignItems: 'center', justifyContent: 'center', height: Dimensions.get('window').height - 150}}>
            <ActivityIndicator size={'large'} color={'#87CEFA'}/>
          </View>
        )}
        {enterprises.map((item, index) => (
          <Card 
          key={index}
          title={item.enterprise_name} 
          subtitle={item.enterprise_type.enterprise_type_name} 
          image={`https://empresas.ioasys.com.br/${item.photo}`}
          onPress={() => navigateToEnterpriseDetails(item)}
        />
        ))}        
        <View style={styles.container}>
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  )
}

export default HomeScreen;
