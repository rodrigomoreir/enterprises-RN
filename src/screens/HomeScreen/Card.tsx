import React, { memo } from 'react'
import { Dimensions } from 'react-native'
import styled from 'styled-components/native'
import themeColors from '../../theme/theme'
import addOpacity from '../../utils/addOpacity'

interface Props {
  title: string
  subtitle: string
  image: string
  onPress: () => void
}

const screenWidth = Dimensions.get('window').width

const Card: React.FC<Props> = ({ title, subtitle, image, onPress }) => {
  return (
    <StyledTouchableOpacity onPress={onPress}>
      <StyledViewImage>
        <StyledImage source={{ uri: image }} />
      </StyledViewImage>
      <StyledViewText>
        <StyledTitle>{title}</StyledTitle>
        <StyledSubtitle>{subtitle}</StyledSubtitle>
      </StyledViewText>
    </StyledTouchableOpacity>
  )
}

const StyledTouchableOpacity = styled.TouchableOpacity`
  justify-content: center;
  height: 250px;
  width: ${screenWidth - 30}px;
  align-items: center;
  margin-top: 10px;
  margin-bottom: 10px;
  border-width: 1px;
  border-color: ${({ theme }) => theme.colors.grey};
  border-radius: 20px;
`
const StyledViewImage = styled.View`
  flex: 1;
  width: 100%;
  height: 100%;
  border-bottom-width: 1px;
  border-bottom-color: ${({ theme }) => theme.colors.grey};
`
const StyledImage = styled.Image`
  height: 100%;
  width: 100%;
  border-top-left-radius: 20px;
  border-top-right-radius: 20px;
`
const StyledViewText = styled.View`
  width: 100%;
  border-bottom-left-radius: 20px;
  border-bottom-right-radius: 20px;
  background-color: ${addOpacity(themeColors.colors.blue, 0.4)};
`
const StyledTitle = styled.Text`
  align-self: center;
  font-size: 20px;
  line-height: 26.38px;
  font-style: normal;
  font-weight: bold;
  color: ${({ theme }) => theme.colors.darkGrey};
`
const StyledSubtitle = styled.Text`
  align-self: center;
  font-size: 15px;
  line-height: 26.38px;
  font-style: normal;
  font-weight: bold;
  color: ${({ theme }) => theme.colors.darkGrey};
`

export default memo(Card)
