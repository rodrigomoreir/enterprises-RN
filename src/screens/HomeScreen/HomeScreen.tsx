import React, { useEffect } from 'react'
import { useNavigation } from '@react-navigation/native'
import { useDispatch, useSelector } from 'react-redux'
import {
  KeyboardAvoidingView,
  Platform,
  ActivityIndicator,
  SafeAreaView,
  FlatList,
} from 'react-native'
import styled from 'styled-components/native'
import { enterpriseActions } from '../../store/slices/enterpriseSlice'
import Card from './Card'
import { Header } from '../../core'
import { TextField } from '../../core'
import { Form } from '@unform/mobile'
import { EnterpriseProps } from './Entities'
import { screenWidth, screenHeight } from '../../utils'

const HomeScreen: React.FC = () => {
  const navigation = useNavigation()

  const dispatch = useDispatch()
  const userData = useSelector(state => state.user.data)

  const {
    data: enterpriseData,
    error: isError,
    isLoading: isLoading,
  } = useSelector(data => data.enterprise)
  const hasData = enterpriseData.length

  const navigateToEnterpriseDetails = (enterprise: EnterpriseProps) => {
    navigation.navigate('EnterpriseDetails', { enterprise })
  }
  const navigateToUserProfile = () => navigation.navigate('UserProfile')

  const searchEnterprise = (filter: string) => {
    const type = parseInt(filter)

    if (type < 30) {
      const filters = {
        headers: userData,
        filter: type,
      }
      dispatch(enterpriseActions.getEnterprisesDataFilteredType(filters))
    }
    const filters = {
      headers: userData,
      filter,
    }
    dispatch(enterpriseActions.getEnterprisesDataFiltered(filters))
  }

  useEffect(() => {
    dispatch(enterpriseActions.getEnterprisesData(userData))
  }, [dispatch, userData])

  console.log('Home rendered')

  return (
    <KeyboardAvoidingView
      style={{ flex: 1 }}
      behavior={Platform.OS === 'ios' ? 'padding' : undefined}
      enabled>
      <SafeAreaView>
        <Header title={'Empresas'} goProfile={navigateToUserProfile} />
        <Form onSubmit={() => { }} style={{ flexDirection: 'row' }}>
          <StyledTextField
            name={'Pesquisar'}
            buttonSearch={() => { }}
            placeholder={'Pesquisar'}
            autoCorrect={false}
            onChangeText={filter => searchEnterprise(filter)}
            onSubmitEditing={() => { }}
          />
        </Form>

        <StyledContainer>
          {isLoading && (
            <StyledLoadingView>
              <ActivityIndicator size={'large'} color={'#87CEFA'} />
            </StyledLoadingView>
          )}
          {!isLoading && !isError && !!hasData && (
            <FlatList
              data={enterpriseData}
              keyExtractor={item => String(item.id)}
              renderItem={({ item }) => (
                <Card
                  title={item.enterprise_name}
                  subtitle={item.enterprise_type.enterprise_type_name}
                  image={`https://empresas.ioasys.com.br/${item.photo}`}
                  onPress={() => navigateToEnterpriseDetails(item)}
                />
              )}
              contentContainerStyle={{ alignItems: 'center' }}
            />
          )}
          {!hasData && !isLoading && !isError && (
            <StyledLoadingView>
              <StyledText>NÃ£o foram encontradas Empresas</StyledText>
            </StyledLoadingView>
          )}
        </StyledContainer>
      </SafeAreaView>
    </KeyboardAvoidingView>
  )
}

const StyledContainer = styled.View`
  width: ${screenWidth}px;
`
const StyledLoadingView = styled.View`
  height: ${screenHeight - 250}px;
  align-items: center;
  justify-content: center;
`
const StyledText = styled.Text`
  font-style: italic;
`
const StyledTextField = styled(TextField)`
  height: 40px;
  border-radius: 20px;
  margin-left: 15px;
  align-self: flex-start;
  width: ${screenWidth - 30}px;
  margin-bottom: 10px;
`
export default HomeScreen
