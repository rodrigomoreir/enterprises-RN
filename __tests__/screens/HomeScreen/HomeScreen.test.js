import React from 'react'
import HomeScreen from '../../../src/screens/HomeScreen'

import {
  renderWithTheme,
  matchSnapshotWithTheme,
} from '../../../src/functions/testing'

describe('Home Screen', () => {
  test('matches snapshot', () => {
    matchSnapshotWithTheme(<HomeScreen />)
  })

  test('renders correctly', () => {
    renderWithTheme(<HomeScreen />)
  })
})
