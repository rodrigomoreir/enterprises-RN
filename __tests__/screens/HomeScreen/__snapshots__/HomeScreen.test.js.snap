// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Screen matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "borderBottomWidth": 2,
            "borderColor": "#a9a9a9",
            "flexDirection": "row",
            "height": 60,
            "justifyContent": "space-between",
            "paddingHorizontal": 10,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "width": 40,
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#696969",
              "fontSize": 20,
              "fontStyle": "normal",
              "fontWeight": "bold",
              "lineHeight": 26.38,
              "textAlign": "center",
            },
          ]
        }
      >
        Empresas
      </Text>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "borderColor": "#69696999",
            "borderRadius": 20,
            "borderWidth": 1,
            "height": 40,
            "justifyContent": "center",
            "opacity": 1,
            "width": 40,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/profile.png",
            }
          }
          style={
            Array [
              Object {
                "borderRadius": 20,
                "maxHeight": 38,
                "maxWidth": 38,
                "tintColor": "#87cefaCC",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "borderColor": "#696969",
              "borderRadius": 10,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 50,
              "marginVertical": 10,
              "paddingHorizontal": 5,
              "width": 670,
            },
            Array [
              Object {
                "alignSelf": "flex-start",
                "borderRadius": 20,
                "height": 40,
                "marginBottom": 10,
                "marginLeft": 15,
                "width": 720,
              },
            ],
            false,
            false,
            ,
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#87cefa99",
              "borderBottomLeftRadius": 20,
              "borderTopLeftRadius": 20,
              "height": 39,
              "justifyContent": "center",
              "left": -6,
              "opacity": 1,
              "width": 55,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/assets/consultas_24px.png",
              }
            }
            style={
              Array [
                Object {
                  "tintColor": "#696969CC",
                },
              ]
            }
          />
        </View>
        <TextInput
          allowFontScaling={true}
          autoCorrect={false}
          defaultValue=""
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onSubmitEditing={[Function]}
          placeholder="Pesquisar"
          rejectResponderTermination={true}
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
              },
            ]
          }
          underlineColorAndroid="transparent"
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "height": 1084,
              "justifyContent": "center",
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontStyle": "italic",
              },
            ]
          }
        >
          NÃ£o foram encontradas Empresas
        </Text>
      </View>
    </View>
  </RCTSafeAreaView>
</View>
`;
